<B> THIS NOTEBOOK IS THE FIRST OF THE CLASSIFICATION NOTEBOOKS MODULE. THE PURPOSE OF THIS NOTEBOOK IS EXPLORATORY DATA ANALYSIS AND CLEANING </B>

In [174]:
from sklearn.datasets import load_iris
from sklearn.neighbors import KNeighborsClassifier
from sklearn import metrics
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
%config InlineBackend.figure_formats = ['svg']  # or 'retina'
%matplotlib inline

sns.set(context='notebook', style='whitegrid')

In [175]:
# Pandas settings to display all columns (there are 31 columsn in this dataset)
pd.set_option("display.max_columns", None)

In [176]:
Asteroid_data_prelim = pd.read_csv('Asteroid_Updated.csv')

/Users/rachelbryn/opt/anaconda3/envs/metis/lib/python3.8/site-packages/IPython/core/interactiveshell.py:3169: DtypeWarning: Columns (0,10,15,16,23,24) have mixed types.Specify dtype option on import or set low_memory=False.
  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,


In [177]:
Asteroid_data_prelim.head()

name         a         e          i          om           w         q  \
0    Ceres  2.769165  0.076009  10.594067   80.305532   73.597694  2.558684   
1   Pallas  2.772466  0.230337  34.836234  173.080063  310.048857  2.133865   
2     Juno  2.669150  0.256942  12.988919  169.852760  248.138626  1.983332   
3    Vesta  2.361418  0.088721   7.141771  103.810804  150.728541  2.151909   
4  Astraea  2.574249  0.191095   5.366988  141.576605  358.687607  2.082324   

         ad     per_y  data_arc condition_code  n_obs_used     H neo pha  \
0  2.979647  4.608202    8822.0              0        1002  3.34   N   N   
1  3.411067  4.616444   72318.0              0        8490  4.13   N   N   
2  3.354967  4.360814   72684.0              0        7104  5.33   N   N   
3  2.570926  3.628837   24288.0              0        9325  3.20   N   N   
4  3.066174  4.130323   63507.0              0        2916  6.85   N   N   

  diameter                 extent  albedo    rot_per       GM     BV     UB  \
0    939.4  964.4 x 964.2 x 891.8  0.0900   9.074170  62.6284  0.713  0.426   
1      545            582x556x500  0.1010   7.813200  14.3000  0.635  0.284   
2  246.596                    NaN  0.2140   7.210000      NaN  0.824  0.433   
3    525.4  572.6 x 557.2 x 446.4  0.4228   5.342128  17.8000  0.782  0.492   
4  106.699                    NaN  0.2740  16.806000      NaN  0.826  0.411   

   IR spec_B spec_T     G     moid class         n          per          ma  
0 NaN      C      G  0.12  1.59478   MBA  0.213885  1683.145708   77.372096  
1 NaN      B      B  0.11  1.23324   MBA  0.213503  1686.155999   59.699133  
2 NaN     Sk      S  0.32  1.03454   MBA  0.226019  1592.787285   34.925016  
3 NaN      V      V  0.32  1.13948   MBA  0.271609  1325.432765   95.861936  
4 NaN      S      S   NaN  1.09589   MBA  0.238632  1508.600458  282.366289

In [178]:
# Rename columns 
Asteroid_data_prelim.rename(columns = {'a' : 'Semi_major_axis_AU','e':'Eccentricity','i':'Inclination'
                                       ,'om':'Longitude_of_asc_node'
                                       ,'w':'Argument_of_perhelion'
                                       ,'q':'Perihelion_distance_AU'
                                       ,'ad':'Aphelion_distance'
                                       ,'per_y':'Orbital_Period_Years'
                                       ,'H':'Absolute_Mag_Parameter'
                                       ,'neo':'Near_Earth_Object'
                                       ,'pha':'Hazardous'
                                       ,'moid':'Earth_Min_Orbital_Intersect_Distance_AU'
                                       ,'n':'Mean_Motion_degrees_per_day'
                                       ,'per':'Orbital_Period_Days'
                                       ,'ma':'Mean_Anomaly_Degrees'}, inplace= True)

In [179]:
# Explore the NaN's and NULL's
Asteroid_data_prelim.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 839714 entries, 0 to 839713
Data columns (total 31 columns):
 #   Column                                   Non-Null Count   Dtype  
---  ------                                   --------------   -----  
 0   name                                     21967 non-null   object 
 1   Semi_major_axis_AU                       839712 non-null  float64
 2   Eccentricity                             839714 non-null  float64
 3   Inclination                              839714 non-null  float64
 4   Longitude_of_asc_node                    839714 non-null  float64
 5   Argument_of_perhelion                    839714 non-null  float64
 6   Perihelion_distance_AU                   839714 non-null  float64
 7   Aphelion_distance                        839708 non-null  float64
 8   Orbital_Period_Years                     839713 non-null  float64
 9   data_arc                                 824240 non-null  float64
 10  condition_code                  

In [180]:
# Get rid of any observations with NaN's or NULL's
Asteroid_data_NaN_Removed= Asteroid_data_prelim.dropna(axis=0, how= "all")

In [181]:
Asteroid_data_NaN_Removed.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 839714 entries, 0 to 839713
Data columns (total 31 columns):
 #   Column                                   Non-Null Count   Dtype  
---  ------                                   --------------   -----  
 0   name                                     21967 non-null   object 
 1   Semi_major_axis_AU                       839712 non-null  float64
 2   Eccentricity                             839714 non-null  float64
 3   Inclination                              839714 non-null  float64
 4   Longitude_of_asc_node                    839714 non-null  float64
 5   Argument_of_perhelion                    839714 non-null  float64
 6   Perihelion_distance_AU                   839714 non-null  float64
 7   Aphelion_distance                        839708 non-null  float64
 8   Orbital_Period_Years                     839713 non-null  float64
 9   data_arc                                 824240 non-null  float64
 10  condition_code                  

In [182]:
# Create a dataframe with only columsn I want to use. 
# Columns with over 830,000 non-null values
Asteroid_data_NaN_Removed.drop(['name','extent','albedo','condition_code','rot_per','GM','BV','UB','IR','spec_B','spec_T','G','Hazardous','Near_Earth_Object','n_obs_used'], axis=1, inplace= True)  


In [183]:
Asteroid_data_NaN_Removed

Semi_major_axis_AU  Eccentricity  Inclination  Longitude_of_asc_node  \
0                 2.769165      0.076009    10.594067              80.305532   
1                 2.772466      0.230337    34.836234             173.080063   
2                 2.669150      0.256942    12.988919             169.852760   
3                 2.361418      0.088721     7.141771             103.810804   
4                 2.574249      0.191095     5.366988             141.576605   
...                    ...           ...          ...                    ...   
839709            2.812945      0.664688     4.695700             183.310012   
839710            2.645238      0.259376    12.574937               1.620020   
839711            2.373137      0.202053     0.732484             176.499082   
839712            2.260404      0.258348     9.661947             204.512448   
839713            2.546442      0.287672     5.356238              70.709555   

        Argument_of_perhelion  Perihelion_distance_AU  Aphelion_distance  \
0                   73.597694                2.558684           2.979647   
1                  310.048857                2.133865           3.411067   
2                  248.138626                1.983332           3.354967   
3                  150.728541                2.151909           2.570926   
4                  358.687607                2.082324           3.066174   
...                       ...                     ...                ...   
839709             234.618352                0.943214           4.682676   
839710             339.568072                1.959126           3.331350   
839711             198.026527                1.893638           2.852636   
839712             148.496988                1.676433           2.844376   
839713             273.483265                1.813901           3.278983   

        Orbital_Period_Years  data_arc  Absolute_Mag_Parameter diameter  \
0                   4.608202    8822.0                   3.340    939.4   
1                   4.616444   72318.0                   4.130      545   
2                   4.360814   72684.0                   5.330  246.596   
3                   3.628837   24288.0                   3.200    525.4   
4                   4.130323   63507.0                   6.850  106.699   
...                      ...       ...                     ...      ...   
839709              4.717914   17298.0                  20.400      NaN   
839710              4.302346      16.0                  17.507      NaN   
839711              3.655884       5.0                  18.071      NaN   
839712              3.398501      10.0                  18.060      NaN   
839713              4.063580      11.0                  17.406      NaN   

        Earth_Min_Orbital_Intersect_Distance_AU class  \
0                                      1.594780   MBA   
1                                      1.233240   MBA   
2                                      1.034540   MBA   
3                                      1.139480   MBA   
4                                      1.095890   MBA   
...                                         ...   ...   
839709                                 0.032397   APO   
839710                                 0.956145   MBA   
839711                                 0.893896   MBA   
839712                                 0.680220   MBA   
839713                                 0.815280   MBA   

        Mean_Motion_degrees_per_day  Orbital_Period_Days  Mean_Anomaly_Degrees  
0                          0.213885          1683.145708             77.372096  
1                          0.213503          1686.155999             59.699133  
2                          0.226019          1592.787285             34.925016  
3                          0.271609          1325.432765             95.861936  
4                          0.238632          1508.600458            282.366289  
...                             ...                  ...      

In [184]:
# Remove columns:
# name
# data_arc
# n_obs_used

Asteroid_data_NaN_Removed.drop(['data_arc'], axis = 1, inplace = True)

In [185]:
Asteroid_data_NaN_Removed.head()

Semi_major_axis_AU  Eccentricity  Inclination  Longitude_of_asc_node  \
0            2.769165      0.076009    10.594067              80.305532   
1            2.772466      0.230337    34.836234             173.080063   
2            2.669150      0.256942    12.988919             169.852760   
3            2.361418      0.088721     7.141771             103.810804   
4            2.574249      0.191095     5.366988             141.576605   

   Argument_of_perhelion  Perihelion_distance_AU  Aphelion_distance  \
0              73.597694                2.558684           2.979647   
1             310.048857                2.133865           3.411067   
2             248.138626                1.983332           3.354967   
3             150.728541                2.151909           2.570926   
4             358.687607                2.082324           3.066174   

   Orbital_Period_Years  Absolute_Mag_Parameter diameter  \
0              4.608202                    3.34    939.4   
1              4.616444                    4.13      545   
2              4.360814                    5.33  246.596   
3              3.628837                    3.20    525.4   
4              4.130323                    6.85  106.699   

   Earth_Min_Orbital_Intersect_Distance_AU class  Mean_Motion_degrees_per_day  \
0                                  1.59478   MBA                     0.213885   
1                                  1.23324   MBA                     0.213503   
2                                  1.03454   MBA                     0.226019   
3                                  1.13948   MBA                     0.271609   
4                                  1.09589   MBA                     0.238632   

   Orbital_Period_Days  Mean_Anomaly_Degrees  
0          1683.145708             77.372096  
1          1686.155999             59.699133  
2          1592.787285             34.925016  
3          1325.432765             95.861936  
4          1508.600458            282.366289

In [186]:
Asteroid_data_NaN_Removed.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 839714 entries, 0 to 839713
Data columns (total 15 columns):
 #   Column                                   Non-Null Count   Dtype  
---  ------                                   --------------   -----  
 0   Semi_major_axis_AU                       839712 non-null  float64
 1   Eccentricity                             839714 non-null  float64
 2   Inclination                              839714 non-null  float64
 3   Longitude_of_asc_node                    839714 non-null  float64
 4   Argument_of_perhelion                    839714 non-null  float64
 5   Perihelion_distance_AU                   839714 non-null  float64
 6   Aphelion_distance                        839708 non-null  float64
 7   Orbital_Period_Years                     839713 non-null  float64
 8   Absolute_Mag_Parameter                   837025 non-null  float64
 9   diameter                                 137636 non-null  object 
 10  Earth_Min_Orbital_Intersect_Dist

In [187]:
# Remove observations where NaN exists in Semi_Major_axis_AU
Asteroid_data_NaN_Removed = Asteroid_data_NaN_Removed.dropna(axis=0, subset=['Semi_major_axis_AU'])

In [188]:
Asteroid_data_NaN_Removed.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 839712 entries, 0 to 839713
Data columns (total 15 columns):
 #   Column                                   Non-Null Count   Dtype  
---  ------                                   --------------   -----  
 0   Semi_major_axis_AU                       839712 non-null  float64
 1   Eccentricity                             839712 non-null  float64
 2   Inclination                              839712 non-null  float64
 3   Longitude_of_asc_node                    839712 non-null  float64
 4   Argument_of_perhelion                    839712 non-null  float64
 5   Perihelion_distance_AU                   839712 non-null  float64
 6   Aphelion_distance                        839708 non-null  float64
 7   Orbital_Period_Years                     839711 non-null  float64
 8   Absolute_Mag_Parameter                   837023 non-null  float64
 9   diameter                                 137636 non-null  object 
 10  Earth_Min_Orbital_Intersect_Dist

In [189]:
# Count the number of nulls in observations under Eccentricity column 
count = Asteroid_data_NaN_Removed["Eccentricity"].count()
print(count)

839712


<B> THE TOTAL NUMBER OF ROWS IS 839,712. LOOKING AT CELL 66 WE SEE THAT FEATURES 0-5 HAVE ALL NON-NULL. WE MUST GET RID OF ROWS WHERE 6-16 HAVE NULL. THIS WILL GIVE US A COMPLETE DATASET TO MODEL. </B> 

In [190]:
# Remove Obzervations where NaN exists in Aphelion_distance
Asteroid_data_NaN_Removed = Asteroid_data_NaN_Removed.dropna(axis=0, subset=['Aphelion_distance'])

In [191]:
Asteroid_data_NaN_Removed.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 839708 entries, 0 to 839713
Data columns (total 15 columns):
 #   Column                                   Non-Null Count   Dtype  
---  ------                                   --------------   -----  
 0   Semi_major_axis_AU                       839708 non-null  float64
 1   Eccentricity                             839708 non-null  float64
 2   Inclination                              839708 non-null  float64
 3   Longitude_of_asc_node                    839708 non-null  float64
 4   Argument_of_perhelion                    839708 non-null  float64
 5   Perihelion_distance_AU                   839708 non-null  float64
 6   Aphelion_distance                        839708 non-null  float64
 7   Orbital_Period_Years                     839708 non-null  float64
 8   Absolute_Mag_Parameter                   837019 non-null  float64
 9   diameter                                 137636 non-null  object 
 10  Earth_Min_Orbital_Intersect_Dist

In [192]:
# Remove Obzervations where NaN exists in Absolute Mag Parameter
Asteroid_data_NaN_Removed = Asteroid_data_NaN_Removed.dropna(axis=0, subset=['Absolute_Mag_Parameter'])

In [193]:
Asteroid_data_NaN_Removed.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 837019 entries, 0 to 839713
Data columns (total 15 columns):
 #   Column                                   Non-Null Count   Dtype  
---  ------                                   --------------   -----  
 0   Semi_major_axis_AU                       837019 non-null  float64
 1   Eccentricity                             837019 non-null  float64
 2   Inclination                              837019 non-null  float64
 3   Longitude_of_asc_node                    837019 non-null  float64
 4   Argument_of_perhelion                    837019 non-null  float64
 5   Perihelion_distance_AU                   837019 non-null  float64
 6   Aphelion_distance                        837019 non-null  float64
 7   Orbital_Period_Years                     837019 non-null  float64
 8   Absolute_Mag_Parameter                   837019 non-null  float64
 9   diameter                                 136889 non-null  object 
 10  Earth_Min_Orbital_Intersect_Dist

In [194]:
# Remove Obzervations where NaN exists Mean_Anomaly_Degrees
Asteroid_data_NaN_Removed = Asteroid_data_NaN_Removed.dropna(axis=0, subset=['Mean_Anomaly_Degrees'])

In [195]:
Asteroid_data_NaN_Removed.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 837013 entries, 0 to 839713
Data columns (total 15 columns):
 #   Column                                   Non-Null Count   Dtype  
---  ------                                   --------------   -----  
 0   Semi_major_axis_AU                       837013 non-null  float64
 1   Eccentricity                             837013 non-null  float64
 2   Inclination                              837013 non-null  float64
 3   Longitude_of_asc_node                    837013 non-null  float64
 4   Argument_of_perhelion                    837013 non-null  float64
 5   Perihelion_distance_AU                   837013 non-null  float64
 6   Aphelion_distance                        837013 non-null  float64
 7   Orbital_Period_Years                     837013 non-null  float64
 8   Absolute_Mag_Parameter                   837013 non-null  float64
 9   diameter                                 136889 non-null  object 
 10  Earth_Min_Orbital_Intersect_Dist

<B> I NEED TO DECIDE WHETHER TO REDUCE MY DATASET TO MEET THE NUMBER OF NON-NULLS IN THE DIAMETER COLUMN: 136,889. I WOULD BE GREATLY REDUCING MY DATASET FOR THE MODEL. HOWEVER, I THINK DIAMETER MAY BE AN IMPORTANT PHYSICAL FEATURE OF THE ASTEROID TO INCLUDE. </B>

In [196]:
# Remove Obzervations where NaN exists in Diameter? 
# Asteroid_data_NaN_Removed = Asteroid_data_NaN_Removed.dropna(axis=0, subset=['diameter'])
# Drop the Diameter column 

In [197]:
Asteroid_data_NaN_Removed.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 837013 entries, 0 to 839713
Data columns (total 15 columns):
 #   Column                                   Non-Null Count   Dtype  
---  ------                                   --------------   -----  
 0   Semi_major_axis_AU                       837013 non-null  float64
 1   Eccentricity                             837013 non-null  float64
 2   Inclination                              837013 non-null  float64
 3   Longitude_of_asc_node                    837013 non-null  float64
 4   Argument_of_perhelion                    837013 non-null  float64
 5   Perihelion_distance_AU                   837013 non-null  float64
 6   Aphelion_distance                        837013 non-null  float64
 7   Orbital_Period_Years                     837013 non-null  float64
 8   Absolute_Mag_Parameter                   837013 non-null  float64
 9   diameter                                 136889 non-null  object 
 10  Earth_Min_Orbital_Intersect_Dist

<B> EXPORT THE CLEANED DATA TO A CSV WHERE I WILL USE IT IN THE NEXT NOTEBOOK: LOGISTICAL REGRESSION </B>

In [201]:
Asteroid_data_NaN_Removed.to_csv('Asteroid_Data_Cleaned_excluding_diameter.csv')

In [202]:
Asteroid_data_NaN_Removed['class'].value_counts()

MBA    745685
OMB     24265
IMB     17438
MCA     16989
APO     11758
AMO      8019
TJN      7371
TNO      3303
ATE      1601
CEN       485
AST        79
IEO        20
Name: class, dtype: int64

In [ ]:
Below is the proportions of the classes when diameter is excluded
MBA    125745
OMB      7455
TJN      1873
IMB       565
APO       485
MCA       336
AMO       268
ATE        95
CEN        48
TNO        13
AST         6